src/Dialect/ONNX/ElementsAttr/ElementsAttrBuilder.hpp

```diff
@@ -244,10 +247,32 @@ class ElementsAttrBuilder {
   // Constructs a transformer that changes every element to the result of
   // applying the given function to the element.
   template <typename Function = WideNum (*)(WideNum)>
-  static inline Transformer functionTransformer(Function fun) {
-    return [fun = std::move(fun)](llvm::MutableArrayRef<WideNum> data) -> void {
-      for (WideNum &n : data)
-        n = fun(n);
+  static inline Transformer functionTransformer(
+      Function fun, mlir::MLIRContext *ctx) {
+    return [fun = std::move(fun), ctx](
+               llvm::MutableArrayRef<WideNum> data) -> void {
+      std::mutex mtx;
+      size_t beginOffset = 0;
```

Comment:
maybe slightly simpler if you omit the mutex and make beginOffset atomic and use beginOffset.fetch_add() to increment it and at the same time read its old value

Reply:
Same comment as above, no need to use a lock to compute the lower and upper bound of the work to be done by a given thread.